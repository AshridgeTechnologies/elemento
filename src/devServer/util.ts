import fs from 'fs'
import axios from 'axios'

export async function getModule(moduleUrl: string, moduleLocalPath: string, moduleName: string) {
    const moduleAlreadyDownloaded = fs.existsSync(moduleLocalPath)
    const moduleDate = moduleAlreadyDownloaded ? fs.statSync(moduleLocalPath).mtime : new Date(0)
    try {
        const resp = await axios.get(moduleUrl, {
            headers: {
                'If-Modified-Since': moduleDate.toUTCString(),
                'Cache-Control': 'max-age=31536000'
            },
            validateStatus: status => status >= 200 && status <= 304
        })
        if (resp.status === 304) {
            console.info(`Latest ${moduleName} already downloaded from`, moduleUrl)
            return true
        }
        if (resp.status !== 200) {
            throw new Error(resp.status + ' ' + resp.statusText)
        }
        const moduleContents = await resp.data
        await fs.promises.writeFile(moduleLocalPath, moduleContents)
        console.info(`Downloaded latest ${moduleName} from`, moduleUrl)
        return true
    } catch (e) {
        console.warn(`Cannot download latest ${moduleName} from`, moduleUrl, (e as Error).message)
        if (moduleAlreadyDownloaded) {
            console.warn(`Using previously downloaded version of ${moduleName}`)
            return true
        } else {
            return false
        }
    }
}