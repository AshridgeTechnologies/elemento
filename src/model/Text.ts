import Element, {ElementType} from './Element.js'

export default class Text implements Element {
    constructor(
        public id: string,
        public name: string,
        public contentExpr: string) {}

    static is(element: Element): element is Text {
        return element.constructor.name === this.name
    }

    kind: ElementType = 'Text'

}
