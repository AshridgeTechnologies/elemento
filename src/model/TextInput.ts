import BaseElement from './BaseElement'
import Element from './Element'
import {ElementId, PropertyValue} from './Types'

export type Properties = {
    readonly initialValue?: PropertyValue,
    readonly maxLength?: PropertyValue,
    readonly multiline?: PropertyValue,
    readonly label?: PropertyValue
}
export default class TextInput extends BaseElement<Properties> implements Element {

    constructor(
        id: ElementId,
        name: string,
        properties: Properties) {
        super(id, name, properties)
    }

    static is(element: Element): element is TextInput {
        return element.constructor.name === this.name
    }

    static defaultValue = ''

    get initialValue() { return this.properties.initialValue }
    get maxLength() { return this.properties.maxLength }
    get multiline() { return this.properties.multiline }
    get label() { return this.properties.label }
}