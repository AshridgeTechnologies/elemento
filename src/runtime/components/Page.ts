import React from 'react'
import {valueOfProps} from '../runtimeFunctions'
import {Stack} from '@mui/material'

type Properties = { path: string, children?: any }

export default function Page({children, path, ...props}: Properties) {
    const propVals = valueOfProps(props)
    return React.createElement(Stack, {
        id: path,
        justifyContent: 'flex-start',
        alignItems: 'flex-start',
        spacing: 2,
        height: '100%', width: '100%',
        sx: {overflowY: 'scroll', padding: 1, position: 'relative'},
        ...propVals,
        children
    })
}
