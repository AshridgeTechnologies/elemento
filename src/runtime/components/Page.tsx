import React from 'react'
import {valueOfProps} from '../runtimeFunctions'
import {Stack} from '@mui/material'

type Properties = {path: string, children?: any}

export default function Page({children, path, ...props}: Properties) {
    const propVals = valueOfProps(props)
    return <Stack id={path}
                  justifyContent="flex-start"
                  alignItems="flex-start"
                  spacing={2}
                  {...propVals}>
        {children}
    </Stack>
}
