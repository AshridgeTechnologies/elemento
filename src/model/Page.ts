import Element, {ElementType} from './Element'
import BaseElement,{equalArrays} from './BaseElement'
import Text from './Text'

export default class Page extends BaseElement implements Element {

    constructor(
        id: string,
        name: string,
        private readonly props: {style?: string},
        public readonly elements: ReadonlyArray<Element>
    ) {
        super(id, name)
    }

    kind: ElementType = 'Page'

    static is(element: Element): element is Page {
        return element.constructor.name === this.name
    }

    protected getElements(): ReadonlyArray<Element> { return this.elements }

    get properties() {return this.props}
    get style() { return this.props.style }

    set(id: string, propertyName: string, value: any): Page {
        if (id === this.id) {
            if (propertyName === 'name') {
                return new Page(this.id, value, this.props, this.elements)
            }
            if (propertyName === 'elements') {
                return new Page(this.id, this.name, this.props, value)
            }

            const updatedProps = {...this.props, [propertyName]:value}
            return new Page(this.id, this.name, updatedProps, this.elements)
        }

        const newElements = this.getElements().map( el => el.set(id, propertyName, value))
        if (!equalArrays(newElements, this.getElements())) {
            return new Page(this.id, this.name, this.props, newElements)
        }

        return this
    }

    static fromJSON({id, name, props, elements}: { id: string; name: string; props: any, elements: any[] }): Page {
        return new Page(id, name, props, elements.map(el => Text.fromJSON(el)))
    }
}