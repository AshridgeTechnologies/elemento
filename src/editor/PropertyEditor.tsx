import React from 'react'
import {startCase} from 'lodash'
import Element from '../model/Element'
import Text from '../model/Text'
import {Box, TextField} from '@mui/material'
import {OnChangeFn} from './Types'
import Page from '../model/Page'
import TextInput from '../model/TextInput'
import UnsupportedValueError from '../util/UnsupportedValueError'

export default function PropertyEditor(props: {element: Element, onChange: OnChangeFn }) {
    const {element, onChange} = props
    const handleChange = (propertyName: string) => (event: any) => {
        const newValue = event.target.value
        onChange(element.id, propertyName, newValue)
    }

    function propertyField<T extends Element>(name: string) {
        return <TextField id={name} label={startCase(name)} variant='outlined' size='small' value={(element as T)[name as keyof T] || ''}
                          onChange={handleChange(name)}/>
    }

    function propertyFields() {

        switch(element.kind) {
            case "App":
                return null

            case "Page":
                return <>
                    {propertyField<Page>('style')}
                </>

            case "Text":
                const text = element as Text
                return <>
                    <TextField id="content" label="Content" variant='outlined' size='small' value={text.contentExpr}
                               onChange={handleChange('contentExpr')}/>
                </>

            case "TextInput":
                return <>
                    {propertyField<TextInput>('initialValue')}
                    {propertyField<TextInput>('maxLength')}
                    {propertyField<TextInput>('label')}
                </>
            default:
                throw new UnsupportedValueError(element.kind)
        }



    }

    const children = propertyFields()
    if (children) {
        return <Box
            component="form"
            sx={{
                '& > :not(style)': { m: 1, width: '25ch' },
            }}
            noValidate
            autoComplete="off"
        >
            <TextField id="id" label="Id" variant='outlined' size='small' value={element.id} disabled/>
            <TextField id="name" label="Name" variant='outlined' size='small' value={element.name}
                       onChange={handleChange('name')}/>
            {children}
        </Box>
    }

    return <div>Unknown element type {element.constructor.name}</div>
}