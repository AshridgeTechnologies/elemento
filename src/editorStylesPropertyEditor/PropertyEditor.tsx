import React, {useState} from 'react'
import Element from '../model/Element'
import {Box, Stack, TextField, TextFieldProps, Typography} from '@mui/material'
import {OnChangeFn} from '../editor/Types'
import PropertyInput from '../editor/PropertyInput'
import {PropertyDef, PropertyType, PropertyValue, StylingProps} from '../model/Types'
import lodash from 'lodash';
import {FILES_ID, TOOLS_ID} from '../model/Project'
import {StylesPropertyEditor} from './StylesPropertyEditor'

const {startCase} = lodash;

function NameTextField(props: TextFieldProps & {readOnly: boolean}) {
    const [changedValue, setChangedValue] = useState<string | undefined>(undefined)
    const onChange = (event: React.ChangeEvent<HTMLInputElement>) => setChangedValue(event.target.value)
    const onKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
        if (event.key === 'Enter') {
            onFinishChange()
        }
    }
    const onFinishChange = () => {
        if (changedValue !== undefined && changedValue !== props.value) {
            props.onChange?.({target: {value: changedValue ?? ''}} as any)
        }
        setChangedValue(undefined)
    }

    return <TextField {...props}
                      variant='outlined' size='small'
                      sx={{flexGrow: 0.7 }}
                      InputProps={{sx: {fontSize: 20}}}
                      inputProps={{readOnly: props.readOnly}}
                      value={changedValue ?? props.value}
                      data-eltype='elementName'
                      label='Name'
                      onChange={onChange}
                      onBlur={onFinishChange}
                      onKeyDown={onKeyDown}
                      helperText={changedValue !== undefined ? 'Renaming - Enter to confirm' : undefined}
    />
}

function NotesTextField(props: TextFieldProps) {
    const [changedValue, setChangedValue] = useState<string | undefined>(undefined)
    const onChange = (event: React.ChangeEvent<HTMLInputElement>) => setChangedValue(event.target.value)
    const onFinishChange = () => {
        if (changedValue !== undefined && changedValue !== props.value) {
            props.onChange?.({target: {value: changedValue ?? ''}} as any)
        }
        setChangedValue(undefined)
    }

    return <TextField {...props}
                      variant='filled' size='small'
                      multiline
                      value={changedValue ?? props.value ?? ''}
                      data-eltype='elementNotes'
                      label='Notes'
                      onChange={onChange}
                      onBlur={onFinishChange}
    />
}

export default function PropertyEditor({element, propertyDefs, onChange, errors = {}}: {element: Element, propertyDefs: PropertyDef[], onChange: OnChangeFn, errors?: object }) {

    function propertyField(name: string, type: PropertyType = 'string', fixedOnly: boolean, readOnly: boolean) {
        const valueFromElement = element[name as keyof object] as unknown as PropertyValue
        const valueFromProps = (element.properties)[name as keyof object] as unknown as PropertyValue
        const propertyValue = readOnly ? valueFromElement : valueFromProps
        const error = errors[name as keyof object]
        const key = `${element.id}.${name}.kind`
        if (type === 'styles') {
            return <StylesPropertyEditor key={key} elementId={element.id} value={propertyValue as StylingProps} onChange={onChange} errors={error}/>
        }
        return <PropertyInput key={key} elementId={element.id} name={name} type={type} value={propertyValue} onChange={onChange} fixedOnly={fixedOnly}
                              readOnly={readOnly} error={error}/>
    }

    const children = propertyDefs.map(({name, type, fixedOnly, readOnly}) => propertyField(name, type, fixedOnly ?? false, readOnly ?? false))
    const readOnly = element.id === FILES_ID || element.id === TOOLS_ID
    const onNameChange = (event: React.ChangeEvent<HTMLInputElement>) => onChange(element.id, 'name', (event.target as HTMLInputElement).value)
    const onNotesChange = (event: React.ChangeEvent<HTMLInputElement>) => onChange(element.id, 'notes', (event.target as HTMLInputElement).value)
    return <Box
        component="form"
        sx={{
            '& > :not(style)': {m: 0.5, width: 'calc(100% - 20px)'},
        }}
        noValidate
        autoComplete="off"
    >
        <Stack direction='row' spacing={2}>
            <NameTextField id='name' value={element.name} readOnly={readOnly} onChange={onNameChange}/>
            <TextField id="formulaName" label="Formula Name" variant='filled' size='small' value={element.codeName} inputProps={{readOnly: true}}
                       sx={{flexGrow: 0.3}}/>
        </Stack>
        <Stack direction='row' spacing={2} alignItems='baseline'>
            <Typography data-testid="elementType" variant='body1'>{startCase(element.kind)}</Typography>
            <Typography data-testid="elementId" variant='body2' title='Elemento internal id for this element'>{element.id}</Typography>
        </Stack>
        <NotesTextField id='notes' value={element.notes} onChange={onNotesChange}/>
        {children}
    </Box>
}

