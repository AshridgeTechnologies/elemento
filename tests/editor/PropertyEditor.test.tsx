/**
 * @jest-environment jsdom
 */
import React, {createElement} from 'react'
import PropertyEditor from '../../src/editor/PropertyEditor'
import Text from '../../src/model/Text'

import {render, screen} from '@testing-library/react'
import Page from '../../src/model/Page'
import TextInput from '../../src/model/TextInput'
import {componentJSON} from '../util/testHelpers'

const onChange = () => {}

const componentProps = (domElement: any) => {
    const propsKey = Object.keys(domElement).find( k => k.startsWith("__reactProps$"))
    return propsKey !== undefined ? domElement[propsKey as string] : null
}

const inputValue = function (label: string) {
    const input = screen.getByLabelText(label) as HTMLInputElement
    return input.value
}

test('PropertyEditor has expected structure for Text', ()=> {
    const element = new Text('id1', 'Text 1', {contentExpr: '"Hi there!"'})
    expect(componentJSON(createElement(PropertyEditor, {element, onChange}))).toMatchSnapshot()
})

test('PropertyEditor has fields for Page', ()=> {
    const element = new Page('id1', 'Page 1', {style: 'funky'}, [])
    render(<PropertyEditor element={element} onChange={onChange}/>)
    expect(inputValue('Name')).toBe('Page 1')
    expect(inputValue('Style')).toBe('funky')
})

test('PropertyEditor shows controlled component for optional fields for Page', ()=> {
    const element = new Page('id1', 'Page 1', {}, [])
    render(<PropertyEditor element={element} onChange={onChange}/>)
    expect(inputValue('Style')).toBe('')
    const styleInput = screen.getByLabelText('Style') as HTMLInputElement
    expect(componentProps(styleInput).value).toBe('')
})

test('PropertyEditor has fields for Text', ()=> {
    const element = new Text('id1', 'Text 1', {contentExpr: '"Hi!"'})
    render(<PropertyEditor element={element} onChange={onChange}/>)
    expect(inputValue('Name')).toBe('Text 1')
    expect(inputValue('Content')).toBe('"Hi!"')
})

test('PropertyEditor has fields for TextInput', ()=> {
    const element = new TextInput('id1', 'Text Input 1', {initialValue: '"Hi!"', maxLength: "10", label: '"Text One"'})
    render(<PropertyEditor element={element} onChange={onChange}/>)
    expect(inputValue('Name')).toBe('Text Input 1')
    expect(inputValue('Label')).toBe('"Text One"')
    expect(inputValue('Initial Value')).toBe('"Hi!"')
    expect(inputValue('Max Length')).toBe('10')
})