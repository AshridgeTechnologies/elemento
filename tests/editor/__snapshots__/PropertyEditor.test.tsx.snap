// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyEditor has expected structure for Text 1`] = `
<form
  autoComplete="off"
  className="MuiBox-root css-xn6ruc"
  noValidate={true}
>
  <div
    className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
  >
    <label
      className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-disabled MuiFormLabel-filled css-1kty9di-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink={true}
      htmlFor="id"
      id="id-label"
    >
      Id
    </label>
    <div
      className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary Mui-disabled MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiOutlinedInput-input MuiInputBase-input Mui-disabled MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
        disabled={true}
        id="id"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="text"
        value="id1"
      />
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-186xcr5"
        >
          <span>
            Id
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
  >
    <label
      className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1kty9di-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink={true}
      htmlFor="name"
      id="name-label"
    >
      Name
    </label>
    <div
      className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
        disabled={false}
        id="name"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="text"
        value="Text 1"
      />
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-186xcr5"
        >
          <span>
            Name
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
  >
    <label
      className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1kty9di-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink={true}
      htmlFor="content"
      id="content-label"
    >
      Content
    </label>
    <div
      className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
        disabled={false}
        id="content"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="text"
        value="\\"Hi there!\\""
      />
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-186xcr5"
        >
          <span>
            Content
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</form>
`;
